EIPS/eip-152.md

```diff
@@ -0,0 +1,232 @@
+
+## Simple Summary
+<!--"If you can't explain it simply, you don't understand it well enough." Provide a simplified and layman-accessible explanation of the EIP.-->
+
+This EIP will enable the Blake2b hash function to run cheaply on the EVM, allowing easier interoperability between Ethereum and Zcash as well as other Equihash-based PoW coins.
+
+## Abstract
+<!--A short (~200 word) description of the technical issue being addressed.-->
+
+This EIP introduces a new precompiled contract which implements the compression function `F` used in the BLAKE2b cryptographic hashing algorithm, for the purpose of allowing interoperability between the EVM and Zcash, as well as introducing more flexible cryptographic hash primitives to the EVM.
+
+## Motivation
+<!--The motivation is critical for EIPs that want to change the Ethereum protocol. It should clearly explain why the existing protocol specification is inadequate to address the problem that the EIP solves. EIP submissions without sufficient motivation may be rejected outright.-->
+
+Besides being a useful cryptographic hash function and SHA3 finalist, BLAKE2b allows for efficient verification of the Equihash PoW used in Zcash, making a BTC Relay - style SPV client possible on Ethereum. A single verification of an Equihash PoW verification requires 512 iterations of the hash function, making verification of Zcash block headers prohibitively expensive if a Solidity implementation of BLAKE2b is used.
+
+The BLAKE2b algorithm is highly optimized for 64-bit CPUs, and is faster than MD5 on modern processors.
+
+Interoperability with Zcash could enable contracts like trustless atomic swaps between the chains, which could provide a much needed aspect of privacy to the very public Ethereum blockchain.
+
+## Specification
+<!--The technical specification should describe the syntax and semantics of any new feature. The specification should be detailed enough to allow competing, interoperable implementations for any of the current Ethereum platforms (go-ethereum, parity, cpp-ethereum, ethereumj, ethereumjs, and [others](https://github.com/ethereum/wiki/wiki/Clients)).-->
+
+We propose adding a precompiled contract at address `0x09` wrapping the [BLAkE2b `F` compression function](https://tools.ietf.org/html/rfc7693#section-3.2).
+
+The precompile requires 6 inputs encoded with `abi.encodePacked` and taking exactly 213 bytes. The encoded inputs are corresponding to the ones specified in the [BLAKE2b RFC](https://tools.ietf.org/html/rfc7693#section-3.2)
+
+- `rounds` - the number of rounds
+- `h` - the state vector
+- `m` - the message block vector
+- `t_0, t_1` - offset counters
+- `f` - the final block indicator flag
+
+```
+[4 bytes for rounds][64 bytes for h][128 bytes for m][8 bytes for t_0][8 bytes for t_1][1 byte for f]
+```
+
+`f` parameter is considered as `true` if it is `nonzero`. For example:
```

Comment:
Could just shorten this: The boolean `f` parameter is considered as `true` if at least 1 bit out of the 8 bits are set.

I know @holiman suggested to relax this to "help Solidity", but Solidity always cleaned booleans in outgoing ABI encoded data (e.g. always setting it to `00000001b`) and the new encoder also enforces that incoming booleans satisfy this requirement.

Reply:
Maybe I am missing something here but isn't 

> The boolean f parameter is considered as true if at least 1 bit out of the 8 bits is set.

the same as 

> `f` parameter is considered as `true` if it is `nonzero`

? 

I know the question was about shortening the sentence and I agree what @axic proposed is easier to reason about but at the same time we are discussing "making this strict" and I think "at least 1 bit out of the 8 bits is set" is not strict 🤔 

I see some benefit in expressing it as "at least 1 bit out of the 8 bits is set". For example,
`00000001` looks as good as `11111111` when representing `true`.

Reply:
My comment had two parts. The first part asking for simplifying the description. The second arguing against the way boolean is handled, but we can take this second part out of the scope of merging this PR.

Reply:
I updated the sentence to say 

> The boolean f parameter is considered as true if at least 1 bit out of the 8 bits is set.

I agree with @axic this is better phrased. 

I did not change the logic about how boolean is handled (the second part of the comment) because it seems like we don't have a consensus here between what @holiman and @axic proposed.

Reply:
Moved this discussion to the "discussion url": https://github.com/ethereum/EIPs/issues/152#issuecomment-508494194

Reply:
See https://github.com/ethereum/EIPs/pull/2129/commits/dcf86f6a60cc19a1352316906a8a83cace2adcc4 for the updated information about strict validation of `f`.